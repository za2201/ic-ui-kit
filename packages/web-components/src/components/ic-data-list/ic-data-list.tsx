import { Component, Element, Host, h, Prop } from "@stencil/core";

@Component({
  tag: "ic-data-list",
  styleUrl: "ic-data-list.css",
  shadow: true,
})
export class DataList {
  @Element() el: HTMLIcDataListElement;

  /**
   * Title for the data list
   */
  @Prop() heading: string;

  /**
   * Set the data heading and all child data rows to small styling
   */
  @Prop() small: boolean = false;

  render() {
    const { heading, small } = this;

    const children = this.el.children;

    if (small) {
      for (let i = 0; i < children.length; i++) {
        children[i].setAttribute("small", "true");
      }
    }

    return (
      <Host class={{ ["small"]: small }}>
        <div class="heading" id="data-list-heading">
          <slot name="heading">
            <ic-typography variant="h3">{heading}</ic-typography>
          </slot>
        </div>
        <div class="divider" />
        <ul aria-labelledby="data-list-heading" class="rows">
          <slot></slot>
        </ul>
      </Host>
    );
  }
}
