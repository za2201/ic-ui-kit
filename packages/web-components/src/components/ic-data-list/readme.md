# ic-data-list



<!-- Auto Generated Below -->


## Properties

| Property  | Attribute | Description                                                   | Type      | Default     |
| --------- | --------- | ------------------------------------------------------------- | --------- | ----------- |
| `heading` | `heading` | Title for the data list                                       | `string`  | `undefined` |
| `small`   | `small`   | Set the data heading and all child data rows to small styling | `boolean` | `false`     |


## Dependencies

### Depends on

- [ic-typography](../ic-typography)

### Graph
```mermaid
graph TD;
  ic-data-list --> ic-typography
  style ic-data-list fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


