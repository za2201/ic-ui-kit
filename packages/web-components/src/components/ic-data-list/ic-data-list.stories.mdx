import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Description,
} from "@storybook/addon-docs";
import { html } from "lit-html";
import readme from "./readme.md";
import DataRow from "../ic-data-row/readme.md";
import Link from "../ic-link/readme.md";
import StatusTag from "../ic-status-tag/readme.md";
import Button from "../ic-button/readme.md";
import TextField from "../ic-text-field/readme.md";
import Typography from "../ic-typography/readme.md";

<Meta title="Web Components/Data list" component="ic-data-list" />

<Description markdown={readme} />

### Default

<Canvas>
  <Story name="Default">
    {html` <ic-data-list heading="My details">
      <ic-data-row label="First name" value="Dave"></ic-data-row>
      <ic-data-row label="Second name" value="Smith"></ic-data-row>
      <ic-data-row label="D.o.B" value="01/01/1990"></ic-data-row>
      <ic-data-row label="Address" value=""
        ><ic-typography variant="body" slot="value"
          >383 Coffee Drive<br />London<br />SW7 988<br />United
          Kingdom</ic-typography
        ></ic-data-row
      >
      <ic-data-row label="Email address" value="dave@dave.com"></ic-data-row>
    </ic-data-list>`}
  </Story>
</Canvas>

### Long text values

<Canvas>
  <Story name="Long text values">
    {html` <ic-data-list
      heading="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua"
    >
      <ic-data-row
        label="Lorem ipsum dolor sit amet, consectetur adipiscing elit"
        value="Dave"
      ></ic-data-row>
      <ic-data-row label="Second name" value="Smith"></ic-data-row>
      <ic-data-row label="D.o.B" value="01/01/1990"></ic-data-row>
      <ic-data-row
        label="Address"
        value="Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat"
        ><ic-link href="/" slot="end-component">Edit</ic-link></ic-data-row
      >
      <ic-data-row label="Second address"
        ><ic-typography variant="body" slot="value"
          >383 Coffee Drive<br />London<br />SW7 988<br />United
          Kingdom</ic-typography
        ></ic-data-row
      >
      <ic-data-row label="Email address" value="dave@dave.com"></ic-data-row>
    </ic-data-list>`}
  </Story>
</Canvas>

### Small

<Canvas>
  <Story name="Small">
    {html` <ic-data-list heading="My details" small>
      <ic-data-row label="First name" value="Dave"></ic-data-row>
      <ic-data-row label="Second name" value="Smith"></ic-data-row>
      <ic-data-row label="D.o.B" value="01/01/1990"></ic-data-row>
      <ic-data-row label="Address"
        ><ic-typography variant="body" slot="value"
          >383 Coffee Drive<br />London<br />SW7 988<br />United
          Kingdom</ic-typography
        ></ic-data-row
      >
      <ic-data-row label="Email address" value="dave@dave.com"></ic-data-row>
    </ic-data-list>`}
  </Story>
</Canvas>

### Link

<Canvas>
  <Story name="Link">
    {html` <ic-data-list heading="My details">
      <ic-data-row label="First name" value="Dave"
        ><ic-link href="/" slot="end-component">Edit</ic-link></ic-data-row
      >
      <ic-data-row label="Second name" value="Smith"
        ><ic-link href="/" slot="end-component">Edit</ic-link></ic-data-row
      >
      <ic-data-row label="D.o.B" value="01/01/1990"
        ><ic-link href="/" slot="end-component">Edit</ic-link></ic-data-row
      >
      <ic-data-row label="Address">
        <ic-typography variant="body" slot="value"
          >383 Coffee Drive<br />London<br />SW7 988<br />United
          Kingdom</ic-typography
        ><ic-link href="/" slot="end-component">Edit</ic-link>
      </ic-data-row>
      <ic-data-row label="Email address" value="dave@dave.com"
        ><ic-link href="/" slot="end-component">Edit</ic-link></ic-data-row
      >
    </ic-data-list>`}
  </Story>
</Canvas>

### Icon button

<Canvas>
  <Story name="Icon button">
    {html` <ic-data-list heading="My details">
      <ic-data-row label="First name" value="Dave"
        ><ic-button variant="icon" aria-label="Edit" slot="end-component"
          ><svg
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.127 22.562l-7.127 1.438 1.438-7.128 5.689 5.69zm1.414-1.414l11.228-11.225-5.69-5.692-11.227 11.227 5.689 5.69zm9.768-21.148l-2.816 2.817 5.691 5.691 2.816-2.819-5.691-5.689z"
            /></svg></ic-button
      ></ic-data-row>
      <ic-data-row label="Second name" value="Smith"
        ><ic-button variant="icon" aria-label="Edit" slot="end-component"
          ><svg
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.127 22.562l-7.127 1.438 1.438-7.128 5.689 5.69zm1.414-1.414l11.228-11.225-5.69-5.692-11.227 11.227 5.689 5.69zm9.768-21.148l-2.816 2.817 5.691 5.691 2.816-2.819-5.691-5.689z"
            /></svg></ic-button
      ></ic-data-row>
      <ic-data-row label="D.o.B" value="01/01/1990"
        ><ic-button variant="icon" aria-label="Edit" slot="end-component"
          ><svg
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.127 22.562l-7.127 1.438 1.438-7.128 5.689 5.69zm1.414-1.414l11.228-11.225-5.69-5.692-11.227 11.227 5.689 5.69zm9.768-21.148l-2.816 2.817 5.691 5.691 2.816-2.819-5.691-5.689z"
            /></svg></ic-button
      ></ic-data-row>
      <ic-data-row label="Address">
        <ic-button variant="icon" aria-label="Edit" slot="end-component"
          ><svg
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.127 22.562l-7.127 1.438 1.438-7.128 5.689 5.69zm1.414-1.414l11.228-11.225-5.69-5.692-11.227 11.227 5.689 5.69zm9.768-21.148l-2.816 2.817 5.691 5.691 2.816-2.819-5.691-5.689z"
            /></svg
        ></ic-button>
        <ic-typography variant="body" slot="value">
          383 Coffee Drive<br />London<br />SW7 988<br />United Kingdom
        </ic-typography>
      </ic-data-row>
      <ic-data-row label="Email address" value="dave@dave.com"
        ><ic-button variant="icon" aria-label="Edit" slot="end-component"
          ><svg
            xmlns="http://www.w3.org/2000/svg"
            width="24"
            height="24"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.127 22.562l-7.127 1.438 1.438-7.128 5.689 5.69zm1.414-1.414l11.228-11.225-5.69-5.692-11.227 11.227 5.689 5.69zm9.768-21.148l-2.816 2.817 5.691 5.691 2.816-2.819-5.691-5.689z"
            /></svg></ic-button
      ></ic-data-row>
    </ic-data-list>`}
  </Story>
</Canvas>

### Adornment

<Canvas>
  <Story name="Adornment">
    {html` <ic-data-list heading="My details">
      <ic-data-row label="First name" value="Dave"
        ><ic-status-tag
          status="success"
          label="complete"
          variant="secondary"
          slot="end-component"
      /></ic-data-row>
      <ic-data-row label="Second name" value="Smith"
        ><ic-status-tag
          status="success"
          label="complete"
          variant="secondary"
          slot="end-component"
      /></ic-data-row>
      <ic-data-row label="D.o.B" value="01/01/1990"
        ><ic-status-tag
          status="success"
          label="complete"
          variant="secondary"
          slot="end-component"
      /></ic-data-row>
      <ic-data-row label="Address">
        <ic-typography variant="body" slot="value">
          383 Coffee Drive<br />London<br />SW7 988<br />United Kingdom
        </ic-typography>
        <ic-status-tag
          status="success"
          label="complete"
          variant="secondary"
          slot="end-component"
        />
      </ic-data-row>
      <ic-data-row label="Email address" value="dave@dave.com"
        ><ic-status-tag
          status="success"
          label="complete"
          variant="secondary"
          slot="end-component"
      /></ic-data-row>
    </ic-data-list>`}
  </Story>
</Canvas>

### Status tag

<Canvas>
  <Story name="Status tag">
    {html` <ic-data-list heading="My details">
      <ic-data-row label="First name"
        ><ic-status-tag
          status="success"
          label="approved"
          variant="secondary"
          slot="value"
      /></ic-data-row>
      <ic-data-row label="Second name"
        ><ic-status-tag
          status="success"
          label="approved"
          variant="secondary"
          slot="value"
      /></ic-data-row>
      <ic-data-row label="D.o.B"
        ><ic-status-tag
          status="success"
          label="approved"
          variant="secondary"
          slot="value"
      /></ic-data-row>
      <ic-data-row label="Address"
        ><ic-status-tag
          status="success"
          label="approved"
          variant="secondary"
          slot="value"
      /></ic-data-row>
      <ic-data-row label="Email address"
        ><ic-status-tag
          status="success"
          label="approved"
          variant="secondary"
          slot="value"
      /></ic-data-row>
    </ic-data-list>`}
  </Story>
</Canvas>

### Editable example

<Canvas>
  <Story name="Editable example">
    {html` <ic-data-list heading="My details">
        <ic-data-row label="First name"
          ><ic-text-field
            class="text-box"
            slot="value"
            label="First name"
            value="Dave"
            readonly
            hide-label
            full-width
          ></ic-text-field
        ></ic-data-row>
        <ic-data-row label="Second name"
          ><ic-text-field
            class="text-box"
            slot="value"
            label="Second name"
            value="Smith"
            readonly
            hide-label
            full-width
          ></ic-text-field
        ></ic-data-row>
        <ic-data-row label="D.o.B"
          ><ic-text-field
            class="text-box"
            slot="value"
            label="D.o.B"
            value="01/01/1990"
            readonly
            hide-label
            full-width
          ></ic-text-field
        ></ic-data-row>
        <ic-data-row label="Address"
          ><ic-text-field
            class="text-box"
            slot="value"
            label="Address"
            value="383 Coffee Drive
London
SW7 988
United Kingdom"
            readonly
            hide-label
            full-width
            rows="4"
          ></ic-text-field
        ></ic-data-row>
        <ic-data-row label="Email address"
          ><ic-text-field
            class="text-box"
            slot="value"
            label="Email address"
            value="dave@dave.com"
            readonly
            hide-label
            full-width
          ></ic-text-field
        ></ic-data-row>
      </ic-data-list>
      <ic-button
        onclick="
      var textFields = document.querySelectorAll('.text-box');
      for (var i=0; i < textFields.length; i++) {
        textFields[i].setAttribute('readonly', 'false');
      }"
        style="margin-right: var(--ic-space-md);margin-top: var(--ic-space-lg);"
        >Edit</ic-button
      >
      <ic-button
        onclick="
      var textFields = document.querySelectorAll('.text-box');
      for (var i=0; i < textFields.length; i++) {
        textFields[i].setAttribute('readonly', 'true');
      }"
        variant="secondary"
        style="margin-top: var(--ic-space-lg);"
        >Confirm</ic-button
      >`}
  </Story>
</Canvas>
