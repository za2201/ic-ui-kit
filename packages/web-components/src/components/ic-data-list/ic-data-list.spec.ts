import { newSpecPage } from "@stencil/core/testing";
import { DataList } from "./ic-data-list";

describe("ic-data-list", () => {
  it("should render with data-row children", async () => {
    const page = await newSpecPage({
      components: [DataList],
      html: `
      <ic-data-list heading="Heading">
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
      </ic-data-list>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with small prop", async () => {
    const page = await newSpecPage({
      components: [DataList],
      html: `
      <ic-data-list heading="Heading" small>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
      </ic-data-list>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with a slotted heading", async () => {
    const page = await newSpecPage({
      components: [DataList],
      html: `
      <ic-data-list>
        <ic-typography variant="h3" slot="heading">Heading</ic-typography>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
        <ic-data-row label="Label" value="value"></ic-data-row>
      </ic-data-list>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render data-row children with slotted content", async () => {
    const page = await newSpecPage({
      components: [DataList],
      html: `
      <ic-data-list heading="Heading">
        <ic-data-row label="Label" value="value"><ic-link href="/" slot="end-component"></ic-link></ic-data-row>
        <ic-data-row label="Label" value="value"><ic-link href="/" slot="end-component"></ic-link></ic-data-row>
        <ic-data-row label="Label" value="value"><ic-link href="/" slot="end-component"></ic-link></ic-data-row>
        <ic-data-row label="Label" value="value"><ic-link href="/" slot="end-component"></ic-link></ic-data-row>
        <ic-data-row label="Label" value="value"><ic-link href="/" slot="end-component"></ic-link></ic-data-row>
        <ic-data-row label="Label" value="value"><ic-link href="/" slot="end-component"></ic-link></ic-data-row>
        <ic-data-row label="Label" value="value"><ic-link href="/" slot="end-component"></ic-link></ic-data-row>
        <ic-data-row label="Label" value="value"><ic-link href="/" slot="end-component"></ic-link></ic-data-row>
      </ic-data-list>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
