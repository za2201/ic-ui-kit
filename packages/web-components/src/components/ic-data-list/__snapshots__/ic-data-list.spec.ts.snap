// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-data-list should render data-row children with slotted content 1`] = `
<ic-data-list heading="Heading">
  <mock:shadow-root>
    <div class="heading" id="data-list-heading">
      <slot name="heading">
        <ic-typography variant="h3">
          Heading
        </ic-typography>
      </slot>
    </div>
    <div class="divider"></div>
    <ul aria-labelledby="data-list-heading" class="rows">
      <slot></slot>
    </ul>
  </mock:shadow-root>
  <ic-data-row label="Label" value="value">
    <ic-link href="/" slot="end-component"></ic-link>
  </ic-data-row>
  <ic-data-row label="Label" value="value">
    <ic-link href="/" slot="end-component"></ic-link>
  </ic-data-row>
  <ic-data-row label="Label" value="value">
    <ic-link href="/" slot="end-component"></ic-link>
  </ic-data-row>
  <ic-data-row label="Label" value="value">
    <ic-link href="/" slot="end-component"></ic-link>
  </ic-data-row>
  <ic-data-row label="Label" value="value">
    <ic-link href="/" slot="end-component"></ic-link>
  </ic-data-row>
  <ic-data-row label="Label" value="value">
    <ic-link href="/" slot="end-component"></ic-link>
  </ic-data-row>
  <ic-data-row label="Label" value="value">
    <ic-link href="/" slot="end-component"></ic-link>
  </ic-data-row>
  <ic-data-row label="Label" value="value">
    <ic-link href="/" slot="end-component"></ic-link>
  </ic-data-row>
</ic-data-list>
`;

exports[`ic-data-list should render with a slotted heading 1`] = `
<ic-data-list>
  <mock:shadow-root>
    <div class="heading" id="data-list-heading">
      <slot name="heading">
        <ic-typography variant="h3"></ic-typography>
      </slot>
    </div>
    <div class="divider"></div>
    <ul aria-labelledby="data-list-heading" class="rows">
      <slot></slot>
    </ul>
  </mock:shadow-root>
  <ic-typography slot="heading" variant="h3">
    Heading
  </ic-typography>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
</ic-data-list>
`;

exports[`ic-data-list should render with data-row children 1`] = `
<ic-data-list heading="Heading">
  <mock:shadow-root>
    <div class="heading" id="data-list-heading">
      <slot name="heading">
        <ic-typography variant="h3">
          Heading
        </ic-typography>
      </slot>
    </div>
    <div class="divider"></div>
    <ul aria-labelledby="data-list-heading" class="rows">
      <slot></slot>
    </ul>
  </mock:shadow-root>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
  <ic-data-row label="Label" value="value"></ic-data-row>
</ic-data-list>
`;

exports[`ic-data-list should render with small prop 1`] = `
<ic-data-list class="small" heading="Heading" small="">
  <mock:shadow-root>
    <div class="heading" id="data-list-heading">
      <slot name="heading">
        <ic-typography variant="h3">
          Heading
        </ic-typography>
      </slot>
    </div>
    <div class="divider"></div>
    <ul aria-labelledby="data-list-heading" class="rows">
      <slot></slot>
    </ul>
  </mock:shadow-root>
  <ic-data-row label="Label" small="true" value="value"></ic-data-row>
  <ic-data-row label="Label" small="true" value="value"></ic-data-row>
  <ic-data-row label="Label" small="true" value="value"></ic-data-row>
</ic-data-list>
`;
